### PR TITLE
:black_circle:  sphere primitive

### PR DESCRIPTION
**Pull request by Quentintnrl**
Add sphere primitives

### CODE REVIEW

File: include/primitives/Properties/SphereProperties.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+#pragma once
+#include <cstddef>
+#include "primitives/Properties/AbstractProperties.hpp"
+#include "json/Json.hpp"
+#include <iostream>
+
+class SphereProperties : public AbstractProperties {
+    public:
+        SphereProperties() : AbstractProperties() { _radius = 1; };
```
Comment (aimericdrk) [inline include/primitives/Properties/SphereProperties.hpp:17]: coding in hpp + _radius = 1.0f

File: shaders/raytracing.frag [hunk 2]
```diff
@@ -135,12 +135,14 @@ Hit ComputeHit(Ray ray) {
         }
 
         for (uint j = mesh.startIdx; j < mesh.endIdx; j++) {
-            Triangle tri = iTriangles[j];
-            Hit hit = RayTriangle(ray, tri);
-            if (hit.hit && hit.distance < closestHit.distance) {
-                closestHit = hit;
-                break;
-            }
+//            if (j == 0) {
```
Comment (aimericdrk) [inline shaders/raytracing.frag:138]: remove comment + if you have to use it you just can do if (!j)

File: src/main.cpp [hunk 3]
```diff
@@ -16,20 +17,25 @@
...
+    else {
+        std::cerr << "Usage: " << argv[0] << " <filename>" << std::endl;
+        return 84;
+    }
+    PropertiesFactory propertiesFactory;
+    ObjectsFactory objectsFactory;
+    propertiesFactory.registerProperties("sphere", [](JsonObject *obj) {
+        return new SphereProperties(obj);
+    });
+    objectsFactory.registerObject("sphere", [](AbstractProperties &properties) -> Object * {
```
Comment (aimericdrk) [inline src/main.cpp:33]: asterix has to be sticked to what's it's pointing to

File: src/primitives/Sphere.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+{
+    _properties.setRadius(radius);
+}
+
+Sphere::Sphere(const Material& material, const Transform& transform, const std::vector<Texture>& textures, const SphereProperties& properties) : Object(material, transform, Sphere::generateVertices(properties), textures)
+{
+    _properties = properties;
+}
+
+Sphere::Sphere(const Sphere &other)  : Object(other) {
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:28]: two space before the :

File: src/primitives/Sphere.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    _properties = other._properties;
+}
+
+Sphere::Sphere(JsonObject *obj)
+{
+    _properties = SphereProperties(obj);
+    _triangles = Sphere::generateVertices(_properties);
+}
+
+Sphere::Sphere(SphereProperties &properties): Object(properties)
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:38]: missing space before :

File: src/primitives/Sphere.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    return _properties;
+}
+
+void Sphere::setProperties(const SphereProperties &properties)
+{
+    _properties = properties;
+}
+
+std::vector<Triangle> Sphere::generateVertices(const SphereProperties &properties) {
+//    std::vector<glm::vec3> vertices;
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:59]: comment inside a function

File: src/primitives/Sphere.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    float r = properties.getRadius();
+    float j = -r + (float)sqrt(2 * pow(r, 2) * 3);
+    float topY = r - j;
+    float bottomY = -topY;
+    float theta = 2 * M_PI / 5;
+    glm::vec3 top = {0, -r, 0};
+    glm::vec3 bottom = {0, r, 0};
+    std::vector<glm::vec3> vertices;
+
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:68]: use 1 line separator instead of 2

File: src/primitives/Sphere.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+void Sphere::setProperties(const SphereProperties &properties)
+{
+    _properties = properties;
+}
+
+std::vector<Triangle> Sphere::generateVertices(const SphereProperties &properties) {
+//    std::vector<glm::vec3> vertices;
+    float r = properties.getRadius();
+    float j = -r + (float)sqrt(2 * pow(r, 2) * 3);
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:61]: sqrt(pow(r, 2) * 6); ?

File: src/primitives/Sphere.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    float topY = r - j;
+    float bottomY = -topY;
+    float theta = 2 * M_PI / 5;
+    glm::vec3 top = {0, -r, 0};
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:65]: 0.0f to force float type i will leave only one comment about this but do it everywhere

File: src/primitives/Sphere.cpp [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+    vertices.push_back(top);
+    for (std::size_t i = 0; i < 5; i++) {
+        auto x = (float)cos(theta * (double)i);
+        auto z = (float)sin(theta * (double)i);
+        vertices.emplace_back(x, topY, z);
+    }
+    vertices.push_back(bottom);
+    for (std::size_t i = 0; i < 5; i++) {
+        auto x = (float)cos(theta * ((double)i + 2 * M_PI / 10));
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:78]: 2 * M_PI / 10  = M_PI / 5

File: src/primitives/Sphere.cpp [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        auto z = (float)sin(theta * ((double)i + 2 * M_PI / 10));
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:79]: same

File: src/primitives/Sphere.cpp [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        vertices.emplace_back(x, bottomY, z);
+    }
+
+    std::vector<Triangle> triangles;
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:83]: remember to declare your vars on the top of the file

File: src/primitives/Sphere.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        triangles.clear();
+        triangles = vertex;
+    }
+    for (auto &triangle: triangles) {
+        triangle.v0 = glm::normalize(triangle.v0) * radius;
+        triangle.v1 = glm::normalize(triangle.v1) * radius;
+        triangle.v2 = glm::normalize(triangle.v2) * radius;
+    }
+    for (auto & t : triangles) {
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:137]: remove space between & and t

File: src/primitives/Sphere.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    _properties = properties;
+}
+
+std::vector<Triangle> Sphere::generateVertices(const SphereProperties &properties) {
+//    std::vector<glm::vec3> vertices;
+    float r = properties.getRadius();
+    float j = -r + (float)sqrt(2 * pow(r, 2) * 3);
+    float topY = r - j;
+    float bottomY = -topY;
+    float theta = 2 * M_PI / 5;
```
Comment (aimericdrk) [inline src/primitives/Sphere.cpp:64]: 2 * M_PI / 5; = M_PI / 2.5;

File: shaders/raytracing.frag [hunk 15]
```diff
@@ -135,12 +135,12 @@ Hit ComputeHit(Ray ray) {
         }
 
         for (uint j = mesh.startIdx; j < mesh.endIdx; j++) {
-            Triangle tri = iTriangles[j];
-            Hit hit = RayTriangle(ray, tri);
-            if (hit.hit && hit.distance < closestHit.distance) {
-                closestHit = hit;
-                break;
-            }
+                Triangle tri = iTriangles[j];
```
Comment (RenardFute) [inline shaders/raytracing.frag:138]: Can you remove this indentation please ?

File: shaders/raytracing.frag [hunk 16]
```diff
@@ -181,7 +181,7 @@ void main() {
     ray.direction = normalize(viewPointWorld - iCameraPosition);
 
     iNumSpheres = 0;
-//    FragColor = vec4(ray.direction, 1.0);
+
     Hit hit = ComputeHit(ray);
-    FragColor = hit.hit ? vec4(hit.normal, 1.0) : vec4(0.0, 0.0, 0.0, 1.0);
+    FragColor = hit.hit ? vec4(hit.normal, 1.0) : vec4(255.0, 255.0, 255.0, 1.0);
```
Comment (RenardFute) [inline shaders/raytracing.frag:186]: can you put it back black ?

File: src/main.cpp
Comment (RenardFute) [inline src/main.cpp:1]: I think you might want to keep the cube in the factory as well as the sphere

**Review by RenardFute — CHANGES_REQUESTED** [review]

**Review by aimericdrk — APPROVED** [review]

**Review by RenardFute — APPROVED** [review]

**Pull request merged**